quickwit/quickwit-indexing/src/source/doc_file_reader.rs

```diff
@@ -0,0 +1,549 @@
+            num_bytes_to_skip,
+        }
+    }
+
+    async fn skip(&mut self) -> io::Result<()> {
+        // allocate on the heap to avoid stack overflows
+        let mut buf = vec![0u8; 64000];
+        while self.num_bytes_to_skip > 0 {
+            let num_bytes_to_read = self.num_bytes_to_skip.min(buf.len());
+            let num_bytes_read = self
+                .reader
+                .read_exact(&mut buf[..num_bytes_to_read])
+                .await?;
+            self.num_bytes_to_skip -= num_bytes_read;
+        }
+        Ok(())
+    }
+
+    /// Reads a line and peeks into the readers buffer. Returns the number of
+    /// bytes read and true the end of the file is reached.
+    async fn read_line_and_peek<'a>(&mut self, buf: &'a mut String) -> io::Result<(usize, bool)> {
+        if self.num_bytes_to_skip > 0 {
+            self.skip().await?;
+        }
+        let line_size = self.reader.read_line(buf).await?;
+        if line_size == 0 {
+            return Ok((0, true));
+        }
+        let next_bytes = self.reader.fill_buf().await?;
+        Ok((line_size, next_bytes.is_empty()))
+    }
+}
+
+pub struct DocFileReader {
+    reader: SkipReader,
+    next_offset: u64,
+}
+
+impl DocFileReader {
+    pub fn empty() -> Self {
+        DocFileReader {
+            reader: SkipReader::new(Box::new(tokio::io::empty()), 0),
+            next_offset: 0,
+        }
+    }
+
+    pub async fn from_uri(
+        storage_resolver: &StorageResolver,
+        uri: &Uri,
+        offset: usize,
+    ) -> anyhow::Result<Self> {
+        let (dir_uri, file_name) = dir_and_filename(uri)?;
+        let storage = storage_resolver.resolve(&dir_uri).await?;
+        let file_size = storage.file_num_bytes(file_name).await?.try_into().unwrap();
+        // If it's a gzip file, we can't seek to a specific offset. `SkipReader`
+        // starts from the beginning of the file, decompresses and skips the
+        // first `offset` bytes.
+        let reader = if uri.extension() == Some("gz") {
+            let stream = storage.get_slice_stream(file_name, 0..file_size).await?;
+            let decompressed_stream = Box::new(GzipDecoder::new(BufReader::new(stream)));
+            DocFileReader {
+                reader: SkipReader::new(decompressed_stream, offset),
+                next_offset: offset as u64,
+            }
+        } else {
+            let stream = storage
+                .get_slice_stream(file_name, offset..file_size)
+                .await?;
+            DocFileReader {
+                reader: SkipReader::new(stream, 0),
+                next_offset: offset as u64,
+            }
+        };
+        Ok(reader)
+    }
+
+    /// Reads the next record from the underlying file. Returns `None` when EOF
+    /// is reached.
+    pub async fn next_record(&mut self) -> anyhow::Result<Option<FileRecord>> {
+        let mut buf = String::new();
+        // TODO add retry if read stream is broken because of a transient error
```

Comment:
can you open a ticket if this is important? We tend to forget TODOs.

Reply:
https://github.com/quickwit-oss/quickwit/issues/5243